### PR TITLE
Gossip protocol implementation.

### PR DESCRIPTION
**Pull request by qnikst**

### CODE REVIEW

File: chain-core/src/property.rs [hunk 1]
```diff
@@ -367,6 +367,13 @@ impl<T: Serialize> Serialize for &T {
     }
 }
 
+pub struct NodeId {
```
Comment (mzabaluev) [inline chain-core/src/property.rs]: Agree with @NicolasDP, this belongs in network-core.

File: chain-core/src/property.rs [hunk 2]
```diff
@@ -367,6 +367,13 @@ impl<T: Serialize> Serialize for &T {
...
+    pub high: u64,
+    pub low: u64,
+}
+
+pub trait Gossip {}
```
Comment (mzabaluev) [inline chain-core/src/property.rs]: This is also a network thing, if it's needed at all.

File: network-core/src/client/gossip.rs [hunk 3]
```diff
@@ -0,0 +1,13 @@
+use super::Error;
+
+use chain_core::property::{Gossip, NodeId};
+
+use futures::prelude::*;
+
+pub trait GossipService<T: Gossip> {
```
Comment (mzabaluev) [inline network-core/src/client/gossip.rs]: Don't use a generic parameter here. `BlockService` was wrong about this, #486 fixes that.

File: network-grpc/src/client.rs [hunk 4]
```diff
@@ -299,6 +299,15 @@ where
     T::deserialize(&mut buf).map_err(|e| core_client::Error::new(core_client::ErrorKind::Format, e))
 }
 
+fn serialize<T>(x: &T) -> Vec<u8>
```
Comment (mzabaluev) [inline network-grpc/src/client.rs]: For better consistency with above, perhaps name it `serialize_to_bytes`.

File: chain-core/src/property.rs (hunk 2 above)
Comment (NicolasDP) [reply chain-core/src/property.rs]: and add `Serialize + Deserialize` too as dependency of this trait

File: network-grpc/src/client.rs (hunk 4 above)
Comment (NicolasDP) [reply network-grpc/src/client.rs]: hence implementing the chain-core's traits : `Serialize + Deserialize` here.

**Comment by qnikst** [comment]
@mzabaluev @NicolasDP Please take another look

File: network-core/src/server/gossip.rs [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+use futures::prelude::*;
+
+use std::{error, fmt};
+
+/// Intreface for the node discovery service implementation
+/// in the p2p network.
+pub trait GossipService {
+    /// Gossip message content.
+    type Message: Gossip + Serialize + Deserialize;
```
Comment (mzabaluev) [inline network-core/src/server/gossip.rs]: `Gossip` already implies `Serialize + Deserialize`.

File: network-core/src/lib.rs [hunk 6]
```diff
@@ -5,11 +5,28 @@
...
 pub mod error;
 
 pub mod client;
 pub mod server;
 
+#[derive(Clone, Debug, PartialEq, PartialOrd)]
+pub struct NodeId(u128);
+
+impl NodeId {
+    pub fn from_bytes(high: u64, low: u64) -> Self {
```
Comment (mzabaluev) [inline network-core/src/lib.rs]: These are not bytes. There may be a `fn from_bytes(bytes: &[u8; 16])`, or more conveniently `fn from_bytes(bytes: &[u8])` with a size assertion.

I'd use the name `from_halves`.

File: network-grpc/src/client.rs [hunk 7]
```diff
@@ -69,12 +72,12 @@ pub mod chain_bounds {
...
 /// make connections and perform requests towards other blockchain nodes.
-pub struct Client<B, S, E> {
+pub struct Client<B, T, S, E> {
     node: gen_client::Node<Connection<S, E, BoxBody>>,
-    _phantom: PhantomData<B>,
+    _phantom: PhantomData<(B, T)>,
 }
 
-impl<B, S, E> Client<B, S, E>
+impl<T, G, S, E> Client<T, G, S, E>
```
Comment (mzabaluev) [inline network-grpc/src/client.rs]: Please use the same names for type parameters of `Client` throughout the source file. It's very confusing to have `T` represent different parameters in different places.

File: network-proto/node.proto [hunk 8]
```diff
@@ -89,6 +89,18 @@ message RecordTransactionResponse {
     bytes id = 2;
 }
 
+
+// Gossip message.
+message GossipMessage {
+    message NodeId {
+      fixed64 high = 1;
+      fixed64 low = 2;
```
Comment (mzabaluev) [inline network-proto/node.proto]: Why do we want to specify `NodeId` in this way, and not with `bytes` of an expected size? Is there any special meaning for the 64-bit halves?

Comment (qnikst) [reply network-proto/node.proto]: I get this idea, but I don't think how we can preserve the current properties we want without major code complication :(

Comment (mzabaluev) [reply network-proto/node.proto]: What are the properties, other than the bit size of the `NodeId`?

**Comment by qnikst** [comment]
I've addressed all the comments and rebased atop the latest version. @mzabaluev please take another look.

File: network-core/src/lib.rs [hunk 9]
```diff
@@ -5,11 +5,37 @@
...
 pub mod error;
 
 pub mod client;
 pub mod server;
 
+#[derive(Clone, Debug, PartialEq, PartialOrd)]
+pub struct NodeId([u8; 16]);
+
+pub enum Error {
+    InvalidSize(usize),
```
Comment (mzabaluev) [inline network-core/src/lib.rs]: This `Error` is now specific to node ID errors, but it is defined on the top level.
Furthermore, there is a mod `error` which is completely unrelated.

Maybe put this stuff along with `NodeId` and `Gossip` to a submodule, tentatively named `gossip`?

File: network-grpc/src/client.rs [hunk 10]
```diff
@@ -335,7 +347,26 @@ where
...
+where
+    T: Gossip,
+{
+    fn from_response(res: gen::node::GossipMessage) -> Result<(NodeId, T), core_client::Error> {
+        let node_id = match res.node_id {
+            None => panic!("no node type"),
+            Some(gen::node::gossip_message::NodeId { content }) => {
+                match NodeId::from_slice(&content) {
+                    Ok(node_id) => node_id,
+                    Err(_v) => panic!("incorrect node encoding"),
```
Comment (mzabaluev) [inline network-grpc/src/client.rs]: We shall not panic from invalid network requests.
Respond with error carrying `Code::InvalidArgument` (to be added in network-core if not already).

File: network-core/src/lib.rs [hunk 11]
```diff
@@ -10,6 +10,36 @@ pub mod error;
 pub mod client;
 pub mod server;
 
+pub mod gossip {
```
Comment (mzabaluev) [inline network-core/src/lib.rs]: That's more like it, though I don't like jamming stuff into `lib.rs`. Copy it out to a separate file, perhaps.
We may also want bikeshed the multiple modules named `gossip` at various levels. How about naming this module `types` and renaming `Error` to `NodeIdError`? Also try making it private, it should not be needed outside the crate.

**Pull request merged**